vpr/src/place/place_constraints.cpp

```diff
@@ -80,39 +80,114 @@ PartitionRegion constrained_macro_locs(const t_pl_macro& pl_macro) {
                 //check that subtile is not an invalid value before changing, otherwise it just stays -1
-                if (block_regions[i].get_sub_tile() != -1) {
-                    modified_reg.set_sub_tile(modified_min_pl_loc.sub_tile);
+                if (block_regions[i].get_sub_tile() != NO_SUBTILE) {
+                    modified_reg.set_sub_tile(block_regions[i].get_sub_tile() - offset.sub_tile);
                 }
 
                 modified_pr.add_to_part_region(modified_reg);
             }
 
             if (num_constrained_members == 1) {
-                macro_pr = modified_pr;
+                macro_head_pr = modified_pr;
             } else {
-                macro_pr = intersection(macro_pr, modified_pr);
+                macro_head_pr = intersection(macro_head_pr, modified_pr);
             }
         }
     }
 
+    //intersect to ensure the head pr does not go outside of grid dimensions
+    macro_head_pr = intersection(macro_head_pr, grid_pr);
+
+    //if the intersection is empty, no way to place macro members together, give an error
+    if (macro_head_pr.empty()) {
+        print_macro_constraint_error(pl_macro);
+    }
+
+    return macro_head_pr;
+}
+
+PartitionRegion update_macro_member_pr(PartitionRegion& head_pr, const t_pl_offset& offset, const PartitionRegion& grid_pr, const t_pl_macro& pl_macro) {
+    std::vector<Region> block_regions = head_pr.get_partition_region();
+    PartitionRegion macro_pr;
+
+    for (unsigned int i = 0; i < block_regions.size(); i++) {
+        Region modified_reg;
+
+        vtr::Rect<int> reg_rect = block_regions[i].get_region_rect();
+
+        modified_reg.set_region_rect(reg_rect.xmin() + offset.x, reg_rect.ymin() + offset.y, reg_rect.xmax() + offset.x, reg_rect.ymax() + offset.y);
+
+        //check that subtile is not an invalid value before changing, otherwise it just stays -1
+        if (block_regions[i].get_sub_tile() != NO_SUBTILE) {
+            modified_reg.set_sub_tile(block_regions[i].get_sub_tile() + offset.sub_tile);
+        }
+
+        macro_pr.add_to_part_region(modified_reg);
+    }
+
+    //intersect to ensure the macro pr does not go outside of grid dimensions
+    macro_pr = intersection(macro_pr, grid_pr);
+
     //if the intersection is empty, no way to place macro members together, give an error
     if (macro_pr.empty()) {
-        VPR_ERROR(VPR_ERROR_PLACE, " \n Feasible floorplanning constraints could not be calculated for the placement macro.\n");
+        print_macro_constraint_error(pl_macro);
     }
 
     return macro_pr;
 }
 
+void print_macro_constraint_error(const t_pl_macro& pl_macro) {
+    VTR_LOG(
+        "Feasible floorplanning constraints could not be calculated for the placement macro. \n"
+        "The placement macro contains the following blocks: \n");
+    for (unsigned int i = 0; i < pl_macro.members.size(); i++) {
+        VTR_LOG("Block #%zu ", size_t(pl_macro.members[i].blk_index));
```

Comment:
Should also print the block name. Ids are OK for developers but hostile for users.
Usually I do that as "Block %s (#%d)" where the %s is the name and the %d is the id.